MetaMask/metamask-extension
# e2e: stabilize Ganache and failOnConsoleError

## **TL;DR**

- Some flaky tests were outputting invisible error messages coming from the Chrome console.  This PR makes them visible.
- If Ganache hit an error while closing, it could cause cascading errors in CI.

## **Details**

### Changes to package.json

- `sprintf-js` was already included by sub-packages, but now it's formally included

### Changes to retry.js

- Test error information is easier to read, with less irrelevant info

### Changes to ganache.js

- Converted to TypeScript
- Catch errors when the Ganache server closes, totally ignore the error if it's `Error: EBUSY: resource busy or locked, unlink` (seems to happen on Windows)

### Changes to run-e2e-test.js

- Log the XML result to the console for more convenient reading on local test runs

### Changes to driver.js

There was a major problem where the function was not working as intended.  If any test didn't declare `failOnConsoleError: false`, a console error would make the test fail, but the error was suppressed and was rather silent.  Instead, it would often cause these cascading errors:
```
Error: Server is already closing or closed.
Error: listen EADDRINUSE: address already in use 127.0.0.1:8545.
```
Because the `failOnConsoleError` was pretty broken, most tests declare `failOnConsoleError: false`, and basically skirt around the issue. It now helpfully logs console errors like this example:
```
----Received an error from Chrome----
TypeError: Cannot read properties of undefined (reading 'length')
  at chrome-extension://pdnpffjhgfnonmboacclhoodligjdnik/ui-9.js:10166:13
  at validate (chrome-extension://pdnpffjhgfnonmboacclhoodligjdnik/ui-9.js:10373:34)
---------End of Chrome error---------
-----failOnConsoleError is true------
```
In the long term, we probably want `failOnConsoleError: true` on most tests, to stop ignoring this.

### Changes to toggle-button.component.js

- Fixed a previous capitalization typo of `data-testId`.  This was causing one of the console errors that used to fail with the `failOnConsoleError` problem in `add-custom-network.spec.js`.

### Changes to networks-form.js

- Fixed two places where `url` could be `undefined` and thus `url.length` threw an error.  This was causing one of the console errors that used to fail with the `failOnConsoleError` problem in `add-custom-network.spec.js`.

## **Manual testing steps**

1. Look at the new CircleCI logs and test passes/failures
2. Look at the new local logs.  If you want to see the new error messages, revert `toggle-button.component.js` and `networks-form.js` one at a time, and run
`yarn test:e2e:single test/e2e/tests/add-custom-network.spec.js --browser chrome --debug`

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
